### PR TITLE
Garage Limit (Ready)

### PR DESCRIPTION
**Pull request by DardoTheMaster**
Vehicle Shop integration ![garlimit]
(https://cloud.githubusercontent.com/assets/14088415/15632278/e868c0fe-258c-11e6-98ba-e5bb6a2f1d95.jpeg)
Impound Menu integration
impoundmenu

# Changelog : Features added
### cfgRemoteExec

- Whitelisted a few functions

### Config Master
- Added 4 parameters for limiting total vehicle amount for each side

### ImpoundAction
- Integration with the limits when a cop tries to impound a vehicle
- New message

### storeVehicle
- Integration with the limits : Select the right parameter basing on playerSide
- New message

### vehicleShopBuy
- Denies player to buy a vehicle if his garage is full.Despite this the player will be able to rent one.
- countVehicles remote call
- canBuy var check with exitWith
- nil set of canBuy var at the end of the script

### unimpound
 - Integration with the limits : Select the right parameter basing on playerSide
- New message

### vehicleGarage 
- None. Dafuq git.

### stringtable
- Messages

### countVehicles
- New function 
- Ability to count vehicles of a specified kind passing the following variables : 
    playerSide,
     Player,
     classname                To determine its kind (Car,Ship,Air)
- Return true/false under canBuy variable.

### getVehicles
- Commented part which ,if uncommented,automatically deletes surplus vehicles.

### spawnVehicle
- Integration with limits
- Check : 
      if the player has more vehicle than the garage limit,then the alive parameter of the spawned vehicle will get changed to '0' on the database, also the dbinfo variable will be set to empty,so the player won't be able to store it . Clean way to 'delete' surplus vehicles.

### vehicleStore

Scheme above

### function.sqf
- Little useful function for adding cash to a remote player

## Features Removed
- None

#756

### CODE REVIEW

File: life_server/Functions/Systems/fn_vehicleStore.sqf [hunk 1]
```diff
@@ -19,7 +19,31 @@ if (count _vInfo > 0) then {
     _plate = _vInfo select 1;
     _uid = _vInfo select 0;
 };
+//count vehicles
```
Comment (Liam0) [inline life_server/Functions/Systems/fn_vehicleStore.sqf]: Any reason to run both queries if vehicle limit is disabled?

**Comment by danielstuart14** [comment]
@Jawshy  @BoGuu
I can't test it right now, can one of you do that for me?

**Comment by Jawshy** [comment]
@DardoTheMaster, could you add the remaining languages to the stringtable. You can leave them blank, use English in the strings as a placeholder, or use Google Translate, until the strings are corrected. 
Should look something like this: 
```xml
        <Key ID="STR_NOTF_TooMuchVehicles">
            <Original>You can't store more than %1 vehicles.</Original>
            <Czech></Czech>
            <French></French>
            <Spanish></Spanish>
            <Italian>Non puoi depositare più di %1 veicoli.</Italian>
            <Polish></Polish>
            <Portuguese></Portuguese>
            <Russian></Russian>
            <German>Du kannst nicht mehr als %1 Fahrzeuge einparken.</German>
        </Key>
```

@danielstuart14, I can test this within the next hour.

**Comment by Jawshy** [comment]
The following messages are written to the local/client.rpt file when the player purchases or rents a vehicle: 
```
12:53:03 Error in expression <ype];
_vehiclesCount = [_query, 2] call DB_fnc_asyncCall select 0;
_vehiclesLimi>
12:53:03   Error position: <DB_fnc_asyncCall select 0;
_vehiclesLimi>
12:53:03   Error Undefined variable in expression: db_fnc_asynccall
12:53:03 File mpmissions\__CUR_MP.Altis\core\shops\fn_vehicleShopBuy.sqf, line 23
12:53:03 Error in expression <sesName];closeDialog 0;};
if (_mode && !_enoughRoom ) exitWith {hint localize "S>
12:53:03   Error position: <_enoughRoom ) exitWith {hint localize "S>
12:53:03   Error Undefined variable in expression: _enoughroom
12:53:03 File mpmissions\__CUR_MP.Altis\core\shops\fn_vehicleShopBuy.sqf, line 72
```

```hpp
    garage_limit = 5; //Allows false and numerical values.
```

I can buy and store a sixth vehicle, but it will almost immediately be removed from my garage without explanation (no hint). 

I would also suggest being able to set the garage_limit for each side. 

As of https://github.com/ArmaLife/Framework/pull/762/commits/e2c436e486e548a90cd98175aacbb6d3b1c73c7b

**Comment by DardoTheMaster** [comment]
@Jawshy  closing and fixing.

**Comment by Jawshy** [comment]
You can continue to commit to this pull request rather than creating a new one. We can just use "squash and merge" when it is ready so it commits cleanly into the master branch.

**Comment by DardoTheMaster** [comment]
So you are able to squash commits before merging?

**Comment by Jawshy** [comment]
Exactly.

**Comment by DardoTheMaster** [comment]
Ok got it

File: Altis_Life.Altis/stringtable.xml [hunk 2]
```diff
@@ -4423,6 +4434,10 @@
             <Portuguese>Multando %1</Portuguese>
             <Polish>Mandat dla %1</Polish>
         </Key>
+        <Key ID="STR_COP_TooMuchVehicles">
```
Comment (Jawshy) [inline Altis_Life.Altis/stringtable.xml]: This key needs the remaining language strings.

File: Altis_Life.Altis/stringtable.xml [hunk 3]
```diff
@@ -1889,6 +1889,7 @@
             <Portuguese>Garagem de Veículos</Portuguese>
             <Polish>Garaż</Polish>
         </Key>
+
```
Comment (Jawshy) [inline Altis_Life.Altis/stringtable.xml]: Delete this line.

**Comment by DardoTheMaster** [comment]
Please , let me know if you have better ways for handling vehicleShopBuy - countVehicles communication.

**Comment by tkcjesse** [comment]
So I'm kind of a lost here. Can you explain in some detail what ya got going on here? Code wise.

**Comment by DardoTheMaster** [comment]
Can i post others schemes?

**Comment by tkcjesse** [comment]
I mean. I just want to know essentially what your objectives/goals were with this. Like what features have you added or removed or adjusted?

**Comment by DardoTheMaster** [comment]
Done. Changelog added.

**Comment by tkcjesse** [comment]
So just curious. For what reason do you/others want to restrict players on how many vehicles they can have in their garage? 

Also, say I am a server admin. I set the max vehicles a player can have out to 3. I pull a boat out and do turtles or something and it sits in the ocean by the other turtle hunters.. I starve and respawn in town. Well shit. I pull out my car (vehicle 2) and drive around. It gets stolen and I die again, I pull out another car. I drive a little ways out of town, bust a tire. Grab a repair kit, go to pull out another car and well shit. I can't... I mean what's the point of limiting how many vehicles can be pulled out by one player at a time? If they want to pull them all out to potentially lose or give to a friend or something why are we limiting them?

**Comment by Jawshy** [comment]
Client/local.rpt file: 
```
13:58:23 "Variable: life_fnc_addcash is not allowed TYPE: CODE NS: MN"
```
Whitelist in [SpyGlass_Config.hpp](https://github.com/ArmaLife/Framework/blob/fb149c892f4199fad0a360403d5c047994bf7db0/Altis_Life.Altis/Config_SpyGlass.hpp). 

Update [remoteExec.txt](https://github.com/ArmaLife/Framework/blob/51fec4672a4b98f9a6da0a5b545ba522a98abadd/BEFilters/remoteExec.txt) to: 
```
5 "" !BIS_fnc_(call|effectKilled(AirDestruction|Secondaries)|execVM) !DB_fnc_(insertRequest|queryRequest|updatePartial|updateRequest) !HC_fnc_(addContainer|addHouse|chopShopSell|deleteDBContainer|getVehicles|insertGang|insertRequest|jailSys|queryRequest|receivekeyofServer) !HC_fnc_(removeGang|sellHouse(Container)?|spawnVehicle|spikeStrip) !HC_fnc_update(Gang|HouseContainers|HouseTrunk|Partial|Request) !HC_fnc_vehicle(Create|Delete|Store|Update) !HC_fnc_wanted(Add|Bounty|Crimes|Fetch|ProfUpdate|Remove) !life_fnc_(addVehicle2Chain|adminid|admininfo|animSync|bountyReceive|broadcast|colorVehicle|copLights) !life_fnc_(copSearch|CopSiren|corpse|demoChargeTimer|flashbang|freezePlayer|gangCreated|gangDisbanded|gangInvite) !life_fnc_(garageRefund|giveDiff|hideObj|impoundMenu|jail(Me|Sys)?) !life_fnc_(jumpFnc|knockedOut|licenseCheck|licensesRead|lightHouse|lockVehicle|medic(Lights|Request|Siren)) !life_fnc_(moveIn|pickupItem|pickupMoney|pulloutVeh|receiveItem|receiveMoney|removeLicenses|restrain|revived) !life_fnc_(robPerson|robReceive|say3D|searchClient|seizeClient|setFuel|simDisable|soundDevice|spikeStripEffect) !life_fnc_(tazeSound|ticketPaid|ticketPrompt|vehicleAnimate) !life_fnc_wanted(Add|Bounty|Crimes|Fetch|Info|List|ProfUpdate|Remove) !life_fnc_wireTransfer !SOCK_fnc_(dataQuery|insertPlayerInfo|updateRequest) !SPY_fnc_(cookieJar|notifyAdmins|observe) !TON_fnc_(addContainer|addHouse|chopShopSell|cleanupRequest|countVehicles) !TON_fnc_clientGang(Kick|Leader|Left) !TON_fnc_(clientGetKey|clientMessage) !TON_fnc_(deleteDBContainer|getID|getVehicles|insertGang|keyManagement|managesc|pickupAction|player_query|recupkeyforHC) !TON_fnc_(removeGang|sellHouse(Container)?|setObjVar|spawnVehicle|spikeStrip) !TON_fnc_update(Gang|HouseContainers|HouseTrunk) !TON_fnc_vehicle(Create|Delete|Store|Update) !="_this call fn_whoDoneit"
```
Difference: 
`!TON_fnc_(addContainer|addHouse|chopShopSell|cleanupRequest)`
to
`!TON_fnc_(addContainer|addHouse|chopShopSell|cleanupRequest|countVehicles)`

As of https://github.com/ArmaLife/Framework/pull/762/commits/8672d82285cd88b50cbe92b69cf40eff986f8cdd.

**Comment by Jawshy** [comment]
The [STR_Shop_Veh_NoRoom](https://github.com/ArmaLife/Framework/pull/762/files#diff-9404bc20cba3745df78ec17e370d7f6aR8394) hint - when trying to purchase a vehicle but not having enough room in the garage - cannot be read in the 3D preview for vehicle shops. I suggest adding `closeDialog 0;` so when a player clicks "buy" they will know why the vehicle is not being purchased. 

Suggest adding the garage limit number in the Config_Master to the garage dialog. This would better explain to players the exact garage limit (if it is specified). 

The following message is written to the server.rpt file when a vehicle is garaged: 
```
2016/06/04, 14:34:32 Error in expression <1' AND active='0' AND type='%2'", _uid, _kind];
_vehiclesCount = [_query,2] call>
2016/06/04, 14:34:32   Error position: <_kind];
_vehiclesCount = [_query,2] call>
2016/06/04, 14:34:32   Error Undefined variable in expression: _kind
2016/06/04, 14:34:32 File life_server\Functions\Systems\fn_vehicleStore.sqf, line 36
```
As of https://github.com/ArmaLife/Framework/pull/762/commits/8672d82285cd88b50cbe92b69cf40eff986f8cdd.

**Comment by Jawshy** [comment]
```hpp
    garage_limit_COP = 2; //Allows false and numerical values.
    garage_limit_OPFOR = false; //Allows false and numerical values.
    garage_limit_MEDIC = 3; //Allows false and numerical values.
    garage_limit_CIVILIAN = 4; //Allows false and numerical values.
```

The civilian limit is working as intended. However, the cop and medic limit are not; there is no limit.

**Comment by DardoTheMaster** [comment]
@tkcjesse you didn't get the point of this pull.
I'm limiting the vehicles that players are able to have in the garage (Type specific,so they'll be able to have : 3 Helis,3 cars and 3 boats),not the vehicles they can pull out,

**Comment by DardoTheMaster** [comment]
@Jawshy  everything should be working. 
I'll need some help for the number on the dialog.

**Comment by tkcjesse** [comment]
Why not just put all vehicles in an array and do a count on the array? Seems a lot simpler.

**Comment by DardoTheMaster** [comment]
Well this is an implementation for the server owners, if you don't want players to have 10+ vehicles in their garage,you can do it by simply editing a parameter.
You can not put vehicles in an array if the vehicles do not exists,that's why i'm using queries

**Comment by tkcjesse** [comment]
What... yes you can. Just do it like houses....

**Comment by DardoTheMaster** [comment]
You are thick uh

Take a look at the spawnvehicle script please, the one called for getting back a vehicle from the garage.

As you can see, the script calls this function
`[_vehicle] remoteExecCall ["life_fnc_addVehicle2Chain",_unit];
`
And,if you read the function,there and only there,the vehicle gets added to your life_vehicles.
Do a try,take a vehicle and impound it,then watch the life_vehicles array from the debug console,it will return an array which contains a null object.

**Comment by tkcjesse** [comment]
I'm talking about making a complete new array. Only for vehicles the player owns.

**Comment by Liam0** [comment]
Just looking over it **quickly** there seem to be a rather unnecessary amount of changes all over the place, why is the function not doing more of the work returning true/false or numeric based on the criterias? simplifying the whole process and reliving people who doesn't want to use vehicleLimit from an excess amount of expressions / variables being ran & defined for no reason.

I might be wrong here but surely there is a better way of going about this.

Why have things like.

> UNCOMMENT HERE TO HAVE SURPLUS VEHICLES AUTOMATICALLY DELETED

just put it into the config file while ur at it?

-fn_impoundAction.sqf
-fn_spawnVehicle.sqf
-canBuy
-life_fnc_addCash

woulden't it be sufficient to use a function in vehiclestore, vehicleshopbuy??

**Pull request closed without merging**